evm-contracts/src/v0.6/dev/Flags.sol

```diff
@@ -0,0 +1,122 @@
+pragma solidity 0.6.6;
+
+
+import "../Owned.sol";
+import "../dev/SimpleAccessControl.sol";
+
+
+/**
+ * @title The Flags contract
+ * @notice Allows flags to signal to any reader on the access control list.
+ * The owner can set flags, or designate other addresses to set flags. The
+ * owner must turn the flags off, other setters cannot.
+ */
+contract Flags is Owned, SimpleAccessControl {
+
+  mapping(address => bool) private flags;
+  mapping(address => bool) private setters;
+
+  event FlagOn(
+    address indexed subject
+  );
+  event FlagOff(
+    address indexed subject
+  );
+  event SetterEnabled(
+    address indexed setter
+  );
+  event SetterDisabled(
+    address indexed setter
+  );
+
+  /**
+   * @notice read the warning flag status of a contract address.
+   * @param subject The contract address being checked for a warning.
+   */
+  function getFlag(address subject)
+    public
+    view
+    checkAccess()
+    returns (bool)
+  {
+    return flags[subject];
+  }
+
+  /**
+   * @notice allows owner to enable the warning flags for mulitple addresses.
+   * @param subjects List of the contract addresses whose flag is being raised
+   */
+  function setFlagsOn(address[] calldata subjects)
+    external
+    returns (bool)
+  {
+    require(msg.sender == owner || setters[msg.sender], "Only callable by enabled setters");
+
+    for (uint256 i = 0; i < subjects.length; i++) {
+      address subject = subjects[i];
+
+      if (!flags[subject]) {
+        flags[subject] = true;
+        emit FlagOn(subject);
+      }
+    }
+  }
+
+  /**
+   * @notice allows owner to disable the warning flags for mulitple addresses.
+   * @param subjects List of the contract addresses whose flag is being lowered
+   */
+  function setFlagsOff(address[] calldata subjects)
+    external
+    onlyOwner()
+    returns (bool)
+  {
+    for (uint256 i = 0; i < subjects.length; i++) {
+      address subject = subjects[i];
+
+      if (flags[subject]) {
+        flags[subject] = false;
+        emit FlagOff(subject);
+      }
+    }
+  }
+
+  /**
+   * @notice allows owner to give other addresses permission to set flags on.
+   * @param added List of the addresses of setters to be enabled.
+   */
+  function enableSetters(address[] calldata added)
```

Comment:
It seems like we are duplicating the logic of the `SimpleAccessControl` here, as we recreated simple access control for the `flags` data. 

It would be cleaner to init `flagsAccessControl = new SimpleAccessControl()` in the constructor, and then use its `hasAccess` function to guard `flags` data access. This would use slightly more gas at the expense of cleaner design and less code, but as it's an `onlyOwner()` modifiable function gas cost should not be a critical concern.

Then we would have two separate logical `SimpleAccessControl` contracts (read access & write access). And we could introduce specific modifiers:

- `_onlyWithWriteAccess` - checked when writing `setFlagsOn`
- `_onlyWithReadAccess` - checked when reading `getFlag`

Reply:
Good call on the duplication, with one catch: SimpleAccessController has a `msg.sender == tx.origin` check that would allow any account to write if the method was not a view method. I've renamed the contract to `SimpleReadAccessController` and created a child of it named `SimpleWriteAccessController` which doesn't have that problem.